File: backend/api/routes/library.py
"""
api/routes/library.py

FastAPI router for the SGF games library — namespaced under
``/library`` so the surface name in the URL matches the
product surface name in the SPA.

Five endpoints, all under ``/library``:

    POST   /library/games/import   — batch import of raw SGFs
    GET    /library/games          — paginated list with sort + filter
    GET    /library/games/{id}     — fetch one game including raw_content
    DELETE /library/games/{id}     — delete one game (cascade-nulls card_source)
    GET    /library/players        — distinct player-name set for autocomplete

Wire-shape contract is recorded in ``docs/notes/sgf-library-plan.md``
and (post-ship) at ``docs/dispatch/backend-to-frontend-sgf-library-status.md``.

Pydantic request/response schemas live at the top of this file
per backend CLAUDE.md (inline at the route until a second consumer
appears). The list-response and detail-response shapes are
structurally identical to the domain ``LibraryGameListItem`` /
``LibraryGame`` value objects, so those are reused directly as
``response_model``; the OpenAPI emit is the same.

License: Public Domain (The Unlicense)
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import (
    get_current_user_id,
    get_db,
    get_game_library_service,
)
from domain.auth import UserId
from domain.errors import BatchTooLargeError
from domain.game_library import (
    GameImportInput,
    GameListFilter,
    GameListSort,
    GameListSortDirection,
    ImportOutcome,
    LibraryGame,
    LibraryGameListItem,
    PlayerCount,
)
from services.game_library_service import GameLibraryService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/library", tags=["library"])


# ─── Request / Response wire shapes ───────────────────────────────────


class ImportGameItem(BaseModel):
    """
    One SGF in a batch import request.

    ``source_path`` is an optional provenance field: the SPA's
    directory-upload UX populates it from
    ``File.webkitRelativePath`` so the user's on-disk organisation
    (e.g., ``sgf_db/1980/1980-09-24.sgf``) survives into
    ``metadata_extra["source_path"]``. Single-file uploads, curl
    clients, and existing scripts can omit the field — the
    backend stores nothing for it then.
    """
    model_config = ConfigDict(frozen=True)
    raw_content: str
    source_path: Optional[str] = None


class ImportGamesRequest(BaseModel):
    """Batch import request body."""
    model_config = ConfigDict(frozen=True)
    games: List[ImportGameItem]


class ImportGamesResponse(BaseModel):
    """
    Batch import response — per-file outcomes in input order.

    Each entry is one of:
      - ``{"status": "created", "game_id": N, "client_game_id": "..."}}``
      - ``{"status": "deduplicated", "game_id": N, "client_game_id": "..."|null}}``
      - ``{"status": "errored", "error": "..."}``
    """
    model_config = ConfigDict(frozen=True)
    outcomes: List[ImportOutcome]


class ListGamesResponse(BaseModel):
    """Paginated list response."""
    model_config = ConfigDict(frozen=True)
    rows: List[LibraryGameListItem]
    total_count: int


class ListPlayersResponse(BaseModel):
    """
    Distinct-player-names + game-count response.

    Each entry pairs a player's name with the number of games it
    appears in (either colour) across the caller's library. The
    list is the deduplicated union of ``player_white`` and
    ``player_black`` values, ordered by descending frequency so
    common players surface first in both the autocomplete dropdown
    and the SPA's two-column player accordion. The SPA fetches
    once on Library tab mount and re-fetches after an import
    completes.
    """
    model_config = ConfigDict(frozen=True)
    players: List[PlayerCount]


# ─── Routes ───────────────────────────────────────────────────────────


@router.post(
    "/games/import",
    response_model=ImportGamesResponse,
    status_code=status.HTTP_200_OK,
)
async def import_games(
    body: ImportGamesRequest,
    service: GameLibraryService = Depends(get_game_library_service),
    db: AsyncSession = Depends(get_db),
    user_id: UserId = Depends(get_current_user_id),
):
    """
    Batch import of SGFs into the caller's games library.

    Per-file outcomes (``created`` / ``deduplicated`` / ``errored``)
    return in input order, regardless of mix — one malformed SGF in
    the batch does not fail the batch (the adapter wraps each row in
    a SAVEPOINT).

    Failure paths surfaced as HTTP errors:

    - 413 ``batch_too_large``: ``len(games)`` exceeds the configured
      per-request cap (``SGF_LIBRARY_IMPORT_BATCH_MAX``). Clients
      with larger collections chunk client-side.
    - 422: malformed request body (Pydantic validation).
    """
    inputs = [
        GameImportInput(
            raw_content=item.raw_content,
            source_path=item.source_path,
        )
        for item in body.games
    ]
    try:
        async with db.begin():
            outcomes = await service.import_games(
                user_id=user_id,
                inputs=inputs,
            )
    except BatchTooLargeError as e:
        raise HTTPException(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            detail={
                "kind": "batch_too_large",
                "detail": str(e),
                "received": e.received,
                "maximum": e.maximum,
            },
        )
    return ImportGamesResponse(outcomes=outcomes)


@router.get("/games", response_model=ListGamesResponse)
async def list_games(
    sort: GameListSort = Query(
        default="created_at",
        description="Column to sort by. Must be in the closed vocabulary.",
    ),
    direction: GameListSortDirection = Query(
        default="desc",
        description="Sort direction.",
    ),
    player_like: Optional[str] = Query(
        default=None,
        description=(
            "Substring match against player_white OR player_black — "
            "any-color player filter. Composes (AND) with the "
            "per-color predicates below."
        ),
    ),
    player_white_like: Optional[str] = Query(default=None),
    player_black_like: Optional[str] = Query(default=None),
    date_from: Optional[str] = Query(default=None),
    date_to: Optional[str] = Query(default=None),
    result_eq: Optional[str] = Query(default=None),
    ruleset_eq: Optional[str] = Query(default=None),
    board_size_eq: Optional[int] = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1),
    service: GameLibraryService = Depends(get_game_library_service),
    user_id: UserId = Depends(get_current_user_id),
):
    """
    Paginated list of the caller's library games plus the total
    match count under the same filter.

    Sort column is validated against the closed Literal vocabulary
    by FastAPI's Pydantic gateway (invalid values → 422); the
    service layer asserts pagination bounds defensively. ``limit``
    is capped at ``SGF_LIBRARY_LIST_LIMIT_MAX`` (currently 500).

    ``raw_content`` is omitted from list rows — fetch via
    ``GET /library/games/{id}`` when needed.
    """
    filt = GameListFilter(
        player_like=player_like,
        player_white_like=player_white_like,
        player_black_like=player_black_like,
        date_from=date_from,
        date_to=date_to,
        result_eq=result_eq,
        ruleset_eq=ruleset_eq,
        board_size_eq=board_size_eq,
    )
    try:
        rows, total = await service.list_games(
            user_id=user_id,
            sort=sort,
            direction=direction,
            filt=filt,
            offset=offset,
            limit=limit,
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(e),
        )
    return ListGamesResponse(rows=rows, total_count=total)


@router.get("/games/{game_id}", response_model=LibraryGame)
async def get_game(
    game_id: int,
    service: GameLibraryService = Depends(get_game_library_service),
    user_id: UserId = Depends(get_current_user_id),
):
    """
    Fetch one library game by id, including ``raw_content``.

    404 if no row exists OR if it exists but belongs to a different
    tenant (404-not-403 invariant).
    """
    game = await service.get_game(user_id=user_id, game_id=game_id)
    if game is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Library game not found",
        )
    return game


@router.delete("/games/{game_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_game(
    game_id: int,
    service: GameLibraryService = Depends(get_game_library_service),
    db: AsyncSession = Depends(get_db),
    user_id: UserId = Depends(get_current_user_id),
):
    """
    Delete one library game by id.

    Cascade behaviour on dependent ``card_source`` rows is
    ``ON DELETE SET NULL`` (the existing schema clause): cards
    minted from this game survive the delete with their
    ``game_source_id`` nulled out. Cards retain their position via
    ``normalized_position_id``; they just lose the source link.

    404 if no row exists OR if it belongs to a different tenant —
    the 404-not-403 invariant preserved with the service's boolean
    return.
    """
    async with db.begin():
        deleted = await service.delete_game(user_id=user_id, game_id=game_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Library game not found",
        )


@router.get("/players", response_model=ListPlayersResponse)
async def list_players(
    service: GameLibraryService = Depends(get_game_library_service),
    user_id: UserId = Depends(get_current_user_id),
):
    """
    Distinct player-name set for the caller's library, ordered by
    descending frequency.

    The SPA fetches this once on Library-tab mount, caches in memory
    (not in the persisted workspace document), and runs autocomplete
    against the in-memory list as the user types into the
    player_white / player_black filter inputs. Re-fetch after an
    import completes.

    Combined (white + black) rather than per-color: the slight
    imprecision — suggesting a name that's only ever been black for
    the player_white filter input — is cosmetic, and the
    implementation is trivially simpler. Cardinality at typical
    library size (~thousands of distinct names) makes the full list
    cheap to ship in one go.
    """
    players = await service.list_players(user_id=user_id)
    return ListPlayersResponse(players=players)

File: frontend/src/types/library.ts
/**
 * src/types/library.ts
 *
 * SGF library domain: camel-case projections of the /library/* wire
 * shapes (list rows, the full game row, filters, import
 * inputs/outcomes, the sort vocabulary). The ACL is
 * `services/library-service.ts`. Carved from the single-file
 * `src/types.ts` (2026-06-10, history-lessons audit §3.15); bodies
 * are verbatim from the pre-split file.
 *
 * License: Public Domain (The Unlicense)
 */

import type { BoardId, GameSourceId, GameDisplayOrdinal } from './ids';

// ── Value Objects (readonly preserved) — SGF library domain ───────────────────
//
// Camel-case domain projections of the /library/* wire shapes
// (LibraryGame, LibraryGameListItem, ListGamesResponse,
// ListPlayersResponse, ImportOutcome, plus the import request items).
// The ACL at `services/library-service.ts` translates between these
// and the generated `components['schemas']['*']` wire types in
// `types/backend.ts`. Branded ids (`GameSourceId`, `BoardId`)
// replace raw `number` / `string` at the boundary; the library row's
// primary key is the same `GameSourceId` the forest navigator uses
// (the library and forest are two views over the same `game_source`
// table), and the row's `client_game_id` UUID is the same `BoardId`
// that boards keyed-by-`client_game_id` use — so opening a library
// game on a board carries the brand through, and a subsequent
// card-mint dedups against the library row via the existing
// `get_or_create_game_source_by_client_id` path.
//
// Per ADR-0008's classification discipline: the sort column is a
// closed Literal union — the camelCase domain names map 1:1 to the
// snake_case wire vocabulary via the ACL. Invalid sort columns are
// 422'd by the backend's Pydantic validator; the frontend's type
// system rules them out at compile time.

export type LibrarySortColumn =
  | 'createdAt'
  | 'date'
  | 'playerWhite'
  | 'playerBlack'
  | 'result'
  | 'ruleset'
  | 'boardSize';

export type LibrarySortDirection = 'asc' | 'desc';

// Distinct-player view row — name + the number of games the player
// appears in across either colour. Backend computes the counts; the
// SPA renders a two-column accordion (name, count) and feeds names
// into the autocomplete suggest. Both surfaces consume the same
// frequency-ordered list.
export interface PlayerCount {
  readonly name: string;
  readonly count: number;
}

// One library row in the list view. Excludes `rawContent` — the SGF
// body ships only via the detail endpoint per the column-projection
// discipline (~2 KB/row × 100 rows would dwarf the metadata).
export interface LibraryGameListItem {
  readonly id: GameSourceId;
  // Per-user-id-enumeration design: no longer `BoardId | null` — the
  // "may be None for legacy rows" exception is closed (every
  // `game_source` row now mints a `client_game_id` at insert time,
  // and the migration backfills historical NULLs). See
  // `.claude/dispatch-reports/per-user-id-enumeration-design.md`,
  // Decision 4's disposition table.
  readonly clientGameId: BoardId;
  readonly playerWhite: string | null;
  readonly playerBlack: string | null;
  readonly date: string | null;
  readonly result: string | null;
  readonly ruleset: string | null;
  readonly boardSize: number | null;
  readonly createdAt: string;  // ISO 8601 — leave as string at the ACL
  // Per-user, gaps-on-delete display ordinal — the library list's
  // row-numbering column. Never round-tripped as a reference.
  readonly displayOrdinal: GameDisplayOrdinal;
}

// Full library row including raw SGF body. Returned by GET
// /library/games/{id}; consumed by the preview pane and the
// "Open in board" flow. `metadataExtra` is the JSON-column blob —
// uppercase SGF property keys (KM, HA, EV, RO, …) plus the lowercase
// `source_path` provenance field stamped at import time.
export interface LibraryGame {
  readonly id: GameSourceId;
  // Per-user-id-enumeration design: see LibraryGameListItem's
  // identical note — no longer `BoardId | null`.
  readonly clientGameId: BoardId;
  readonly playerWhite: string | null;
  readonly playerBlack: string | null;
  readonly date: string | null;
  readonly result: string | null;
  readonly ruleset: string | null;
  readonly boardSize: number | null;
  readonly metadataExtra: Readonly<Record<string, unknown>>;
  readonly createdAt: string;
  readonly rawContent: string;
  readonly displayOrdinal: GameDisplayOrdinal;
}

// Filter predicates for GET /library/games. All optional; omitted
// fields don't constrain the query. Substring-match (`*Like`) on
// player names, lexicographic range on date string, exact match on
// result / ruleset / boardSize.
//
// Per ADR-0001 (state containers drop `readonly`): this is a
// reactive state container — `useLibraryQuery` holds it via
// Vue's `reactive(...)` and the SPA's filter inputs v-model
// fields on it directly to trigger refetches. Keep mutability
// honest at the type level.
export interface LibraryFilter {
  // Any-color player filter — ORs across player_white / player_black
  // on the backend. The "show me all of X's games regardless of
  // colour" affordance, distinct from the per-color filters below
  // which target one side specifically. All three filters AND
  // together when set simultaneously.
  playerLike: string | null;
  playerWhiteLike: string | null;
  playerBlackLike: string | null;
  dateFrom: string | null;
  dateTo: string | null;
  resultEq: string | null;
  rulesetEq: string | null;
  boardSizeEq: number | null;
}

// Per-file input to the batch-import endpoint. `sourcePath` carries
// the directory-upload `File.webkitRelativePath` so the user's
// on-disk layout (`sgf_db/1996/cho-vs-lee.sgf`) survives into
// `metadataExtra.source_path` at the backend. `null` for single-file
// uploads and curl clients.
export interface LibraryImportInput {
  readonly rawContent: string;
  readonly sourcePath: string | null;
}

// Per-file outcome of a batch import. Discriminated union; the
// `status` field is the dispatch witness. `errored` carries the
// per-file failure message (malformed SGF, adapter SAVEPOINT
// failure, etc.) — the batch as a whole stays 200 OK and the
// remaining files are unaffected. `deduplicated.clientGameId` may
// be `null` for legacy rows that pre-date the dedup arc.
// Per-user-id-enumeration design: `deduplicated.clientGameId` is no
// longer `BoardId | null` — the legacy exception is closed the same
// way as `LibraryGameListItem.clientGameId` above.
export type LibraryImportOutcome =
  | { readonly status: 'created'; readonly gameId: GameSourceId; readonly clientGameId: BoardId; readonly displayOrdinal: GameDisplayOrdinal }
  | { readonly status: 'deduplicated'; readonly gameId: GameSourceId; readonly clientGameId: BoardId; readonly displayOrdinal: GameDisplayOrdinal }
  | { readonly status: 'errored'; readonly error: string };

// A parsed SGF file staged by the setup wizard's optional import step,
// not yet sent to the backend (commission rows 1404/1407/1464/1468:
// the wizard's import step is a PURE STAGING core — no backend/DB
// mutation before wizard finish — with the actual upload living in an
// imperative shell that commits only when the wizard finishes;
// cancelling/closing the wizard simply never calls that shell, so the
// staged plan evaporates with zero externally-visible effect). Owned
// by the wizard session (`useSetupWizard.ts`), not by
// `useLibraryImport.ts` — that composable stays the EFFECTFUL path the
// Library tab's own import UI uses unchanged. `fileName` is
// display-only (progress/list rendering); `input` is the exact wire
// payload `LibraryService.importGames` sends when the plan commits.
export interface StagedImportFile {
  readonly fileName: string;
  readonly input: LibraryImportInput;
}
